File: cli/src/commands/ls.ts
import { Command } from "commander";
import * as path from "path";
import chalk from "chalk";
import { findACSRoot, loadProject } from "../utils/project";

export function lsCommand(): Command {
  const cmd = new Command("ls");
  cmd
    .description("List all discovered ACS layers, skills, commands, and agents")
    .option("-p, --path <dir>", "Path to project root (default: cwd)")
    .action((opts: { path?: string }) => {
      const start = opts.path ? path.resolve(opts.path) : process.cwd();
      const root = findACSRoot(start);

      if (!root) {
        console.error(
          chalk.red("✗ No .agents/main.yaml found in this directory or any parent.")
        );
        process.exit(1);
      }

      const project = loadProject(start);
      if (!project) {
        console.error(chalk.red("✗ Failed to load project."));
        process.exit(1);
      }

      const {
        manifest,
        skills,
        contextFiles,
        commands,
        agents,
        workflows,
        hooks,
        profiles,
        toolsets,
        tasks,
        memories,
      } = project;

      console.log(
        `${chalk.bold(manifest.project.name)}  ${chalk.dim(`v${manifest.version}`)}`
      );
      console.log(chalk.dim(manifest.project.description));
      if (manifest.project.language || manifest.project.framework) {
        const lang = [manifest.project.language, manifest.project.framework]
          .filter(Boolean)
          .join(" / ");
        console.log(chalk.dim(lang));
      }
      console.log("");

      if (skills.length > 0) {
        console.log(chalk.bold("Skills"));
        for (const s of skills) {
          console.log(`  ${chalk.cyan(s.name)}`);
          if (s.description) {
            console.log(`    ${chalk.dim(s.description)}`);
          }
        }
        console.log("");
      }

      if (commands.length > 0) {
        console.log(chalk.bold("Commands"));
        for (const c of commands) {
          console.log(`  ${chalk.cyan(path.basename(c, ".md"))}`);
        }
        console.log("");
      }

      if (agents.length > 0) {
        console.log(chalk.bold("Agents"));
        for (const a of agents) {
          console.log(`  ${chalk.cyan(path.basename(a, ".md"))}`);
        }
        console.log("");
      }

      if (workflows.length > 0) {
        console.log(chalk.bold("Workflows"));
        for (const workflow of workflows) {
          console.log(`  ${chalk.cyan(workflow.name)}`);
        }
        console.log("");
      }

      if (hooks.length > 0) {
        console.log(chalk.bold("Hooks"));
        for (const hook of hooks) {
          console.log(`  ${chalk.cyan(hook.name)}`);
        }
        console.log("");
      }

      if (profiles.length > 0) {
        console.log(chalk.bold("Profiles"));
        for (const profile of profiles) {
          console.log(`  ${chalk.cyan(profile.name)}`);
        }
        console.log("");
      }

      if (toolsets.length > 0) {
        console.log(chalk.bold("Tools"));
        for (const toolset of toolsets) {
          console.log(`  ${chalk.cyan(toolset.name)}`);
        }
        console.log("");
      }

      if (tasks.length > 0) {
        console.log(chalk.bold("Tasks"));
        for (const task of tasks) {
          console.log(`  ${chalk.cyan(task.name)}`);
        }
        console.log("");
      }

      if (memories.length > 0) {
        console.log(chalk.bold("Memories"));
        for (const memory of memories) {
          console.log(`  ${chalk.cyan(memory.name)}`);
        }
        console.log("");
      }

      if (contextFiles.length > 0) {
        console.log(chalk.bold("Context"));
        for (const c of contextFiles) {
          console.log(`  ${chalk.dim(path.basename(c))}`);
        }
        console.log("");
      }

      if (
        skills.length === 0 &&
        commands.length === 0 &&
        agents.length === 0 &&
        workflows.length === 0 &&
        hooks.length === 0 &&
        profiles.length === 0 &&
        toolsets.length === 0 &&
        tasks.length === 0 &&
        memories.length === 0 &&
        contextFiles.length === 0
      ) {
        console.log(
          chalk.dim("No layers discovered yet. Run acs init to scaffold them.")
        );
      }
    });

  return cmd;
}

File: packages/python/acs_cli/commands/ls.py
"""acs ls — list all discovered ACS layers, skills, commands, and agents."""

import os
import sys

_HERE = os.path.dirname(__file__)
_REPO_ROOT = os.path.abspath(os.path.join(_HERE, "..", "..", "..", ".."))
_REFIMPL = os.path.join(_REPO_ROOT, "reference-impl", "python")
if _REFIMPL not in sys.path:
    sys.path.insert(0, _REFIMPL)

from acs_parser import find_acs_root, load_project  # noqa: E402


def run(args) -> int:
    start = args.path if args.path else os.getcwd()
    root = find_acs_root(start)

    if root is None:
        print("\033[31m✗ No .agents/main.yaml found in this directory or any parent.\033[0m", file=sys.stderr)
        return 1

    project = load_project(start)
    if project is None:
        print("\033[31m✗ Failed to load project.\033[0m", file=sys.stderr)
        return 1

    manifest = project.manifest
    print(f"\033[1m{manifest.project['name']}\033[0m  \033[2mv{manifest.version}\033[0m")
    print(f"\033[2m{manifest.project.get('description', '')}\033[0m")
    lang = " / ".join(filter(None, [manifest.project.get("language"), manifest.project.get("framework")]))
    if lang:
        print(f"\033[2m{lang}\033[0m")
    print()

    if project.skills:
        print("\033[1mSkills\033[0m")
        for s in project.skills:
            print(f"  \033[36m{s.name}\033[0m")
            if s.description:
                print(f"    \033[2m{s.description}\033[0m")
        print()

    if project.commands:
        print("\033[1mCommands\033[0m")
        for c in project.commands:
            print(f"  \033[36m{os.path.splitext(os.path.basename(c))[0]}\033[0m")
        print()

    if project.agents:
        print("\033[1mAgents\033[0m")
        for a in project.agents:
            print(f"  \033[36m{os.path.splitext(os.path.basename(a))[0]}\033[0m")
        print()

    if project.context_files:
        print("\033[1mContext\033[0m")
        for c in project.context_files:
            print(f"  \033[2m{os.path.basename(c)}\033[0m")
        print()

    if not any([project.skills, project.commands, project.agents, project.context_files]):
        print("\033[2mNo layers discovered yet. Run acs init to scaffold them.\033[0m")

    return 0
